### PR TITLE
Add build and release automation

### PR DESCRIPTION
**Pull request by austin-barrington**
This PR joins the build process for windows and linux into the same build actions. this allows a single action to hold both compiled binaries from that process. They are kept by default for 90 days by github. This has been tested in the PR.

The second part is to build and add a new release with those build assets when a tag is pushed or created on the repository. This is harder to test as it'll only happen when the tag is created.

Note: ~I still need to add the token for GITHUB_TOKEN and I am working out where to get this from.~ the Actions process provides this for us

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      - uses: actions/download-artifact@v2
+        with:
+          name: linux-qstat
+      - uses: "marvinpinto/action-automatic-releases@latest"
+        with:
+          repo_token: "${{ secrets.GITHUB_TOKEN }}"
+          prerelease: false
+          files: |
+            linux-qstat
+            windows-qstat.exe
```
Comment (aaronjpitty) [inline .github/workflows/release.yml:58]: Missing a new line :D

**Review by aaronjpitty — APPROVED** [review]

**Review by danmrichards — APPROVED** [review]

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/download-artifact@v2
+        with:
+          name: windows-qstat.exe
+      - uses: actions/download-artifact@v2
+        with:
+          name: linux-qstat
```
[the remark this answers is not in the document]
Comment (stevenh) [reply .github/workflows/release.yml:53]: Just use path to download all assets, this should result in
```shell
bin/windows_amd64/qstat.exe
bin/linux_amd64/qstat
```

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      - uses: "marvinpinto/action-automatic-releases@latest"
+        with:
+          repo_token: "${{ secrets.GITHUB_TOKEN }}"
+          prerelease: false
+          files: |
+            linux-qstat
+            windows-qstat.exe
```
Comment (stevenh) [inline .github/workflows/release.yml:60]:
```suggestion
            bin/*
            qstat.cfg
            contrib.cfg
            LICENSE.txt
```

Comment (stevenh) [reply .github/workflows/release.yml:60]: All binaries, cfg files and license

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ilammy/msvc-dev-cmd@v1
+      - name: nmake
+        run: |
+          nmake -f Makefile.noauto windows windows_debug
+      - uses: actions/upload-artifact@v2
+        with:
+          name: windows-qstat.exe
+          path: qstat.exe
```
Comment (stevenh) [inline .github/workflows/build.yml:42]:
```suggestion
```
No need to upload the build for PRs

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        run: ./configure
+      - name: make
+        run: make
+      - name: make check
+        run: make check
+      - name: make distcheck
+        run: make distcheck
+      - uses: actions/upload-artifact@v2
+        with:
+          name: linux64-qstat
```
Comment (stevenh) [inline .github/workflows/release.yml:26]:
```suggestion
          name: linux_amd64
```
Standardised naming that also supports bulk downloads

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    runs-on: 	windows-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ilammy/msvc-dev-cmd@v1
+      - name: nmake
+        run: |
+          nmake -f Makefile.noauto windows windows_debug
+      - uses: actions/upload-artifact@v2
+        with:
+          name: windows-qstat.exe
```
Comment (stevenh) [inline .github/workflows/release.yml:39]:
```suggestion
          name: windows_amd64
```
Standardised naming which also supports bulk downloads

**Review by stevenh — CHANGES_REQUESTED** [review]
I would recommend merging #105 which adds binary versioning first then rebase so the Build CI changes aren't needed.

I've also suggested some changes to the layout.

The only other change that might be good is to only run the windows release build separately thus avoiding storing the linux assets.

**Comment by austin-barrington** [comment]
Merged #105 and tomorrow (hopefully) i'll get to making the changes you've suggested above!

File: .github/workflows/release.yaml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+          name: linux_amd64
+          path: qstat
+  build_windows:
+    runs-on: 	windows-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ilammy/msvc-dev-cmd@v1
+      - name: nmake
+        run: |
+          set QSTAT_VERSION=${{ github.event.release.tag_name }} && nmake -f Makefile.noauto windows windows_debug
```
Comment (stevenh) [inline .github/workflows/release.yaml:30]:
```suggestion
          set QSTAT_VERSION=${{ github.event.release.tag_name }} && nmake -f Makefile.noauto windows_debug
```

Comment (stevenh) [reply .github/workflows/release.yaml:30]: No need to build both versions as they just overwrite each other.

File: .github/workflows/release.yaml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          prerelease: false
+          tag_name:  "${{ github.event.release.tag_name }}"
+          files: |
+            binaries.zip
+            qstat.cfg
+            contrib.cfg
+            LICENSE.txt
```
Comment (stevenh) [inline .github/workflows/release.yaml:60]: Missing new line

File: .github/workflows/release.yaml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        run: zip -r binaries.zip bin/*
+      - name: Release
+        uses: "softprops/action-gh-release@v1"
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          prerelease: false
+          tag_name:  "${{ github.event.release.tag_name }}"
+          files: |
+            binaries.zip
```
Comment (stevenh) [inline .github/workflows/release.yaml:57]: Why zip them?

**Review by stevenh — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yaml [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ilammy/msvc-dev-cmd@v1
+      - name: nmake
+        run: |
+          set QSTAT_VERSION=${{ github.event.release.tag_name }} && nmake -f Makefile.noauto windows windows_debug
+      - uses: actions/upload-artifact@v2
+        with:
+          name: windows_amd64
+          path: qstat.exe
```
Comment (stevenh) [inline .github/workflows/release.yaml:34]:
```suggestion
          path: |
            qstat.exe
            qstat.pdb
```

**Review by stevenh — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yaml [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  build_windows:
+    runs-on: 	windows-latest
+    env:
+      QSTAT_VERSION=${{ github.event.release.tag_name }}
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ilammy/msvc-dev-cmd@v1
+      - name: nmake
+        run: |
+          nmake -f Makefile.noauto windows windows_debug
```
Comment (stevenh) [inline .github/workflows/release.yaml:32]:
```suggestion
          nmake -f Makefile.noauto windows_debug
```

**Review by stevenh — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yaml (hunk 9 above)
Comment (austin-barrington) [reply .github/workflows/release.yaml:61]: It makes it easier to upload them as assets and it keeps the dir structure in the zip folder else we have to specify each bin individually. 

When adding new items to this list or other lists. It just means it's handled by the zipping process.
```
          name: windows_amd64
          path: |
            qstat.exe
            qstat.pdb
 ``` 
 
 I don't mind either way but I feel like it's cleaner.

Comment (stevenh) [reply .github/workflows/release.yaml:63]: From a users perspective its another step though isn't it?

Comment (stevenh) [reply .github/workflows/release.yaml:63]: Also the binaries aren't fully useable without the config files, so keep that in mind.

File: .github/workflows/release.yaml [hunk 12]
```diff
@@ -0,0 +1,66 @@
+name: Tag Release
+
+on: 
+  push:
+    tags:
+    - 'v*'
```
Comment (stevenh) [inline .github/workflows/release.yaml:6]: This looks wrong, surely you want it on release not push otherwise you'll be creating assets all the time.

**Review by stevenh — CHANGES_REQUESTED** [review]

**Review by stevenh — APPROVED** [review]
LGTM

**Pull request merged**